Use "Literal" types to implement arguments that only allow certain values

tmke8 opened the pull request:
With this PR, this code:

```python
class MyArgs(Tap):
   planet: Literal["mercury", "venus", "earth", "mars"] = "mars"
```

is equivalent to

```python
parser.add_argument('--planet', type=str, choices=["mercury", "venus", "earth", "mars"], default="mars")
```

The `Literal` type was [introduced in Python 3.8](https://www.python.org/dev/peps/pep-0586/) (see also [here](https://mypy.readthedocs.io/en/latest/literal_types.html) for more information) but is also available in previous versions of Python via the (official) [typing_extensions](https://github.com/python/typing/tree/master/typing_extensions) package. The type works like this:

```python
def f(a: Literal[2, "hi", 3.7]) -> None:
    print(a)

f(2)  # OK
f("ho")  # type checker will show an error (but code will run anyway)
```

So it's a perfect fit for the `choices` argument of `argparse`.

The current implementation only accepts *string* literals (i.e. `Literal[2, 3]` is not allowed) and throws an error if you try to use other literals. This could be changed but would make the code much more complicated.

This PR unfortunately introduces two dependencies: the `typing_extensions` package for the `Literal` type and the [typing_inspect](https://github.com/ilevkivskyi/typing_inspect) package (by one of the authors of the Python `typing` module) for identifying `Literal` types and extracting the values. One reason why I'm using `typing_inspect` is that the way to identify `Literal` types has changed from Python 3.6 to Python 3.7 and that package takes care of all the ugly details. It has convenient functions like `is_literal_type()`.

(Two of the functions that I use from `typing_inspect`, `get_args` and `get_origin`, have already [made it into Python 3.8](https://docs.python.org/3/library/typing.html#typing.get_origin).)

tmke8 committed:
Add string literals to the allowed types of arguments

The types look like this: `Literal["mercury", "venus", "mars"]`.
This means the variable may only take on the value of one of the
strings in the list. The `Literal[...]` can also be wrapped in an
`Optional`, a `List` or a `Set`.

tmke8 committed:
Add tests for the Literal types

martinjm97 commented:
Really interesting PR! I didn't know about the `Literal` type. It's been great to see how you've been extending Tap. @swansonk14  and I will try to look into it soon, but it might be a few weeks. Thanks again!

tmke8 commented:
No problem! I really like the library.

I have replaced the use of `argparse` in our codebase with Tap and found a few things that it couldn't do yet, so I added those features in my fork. Now I'm starting to send PRs with those changes to you to see if you like them.

There are a few more changes in my fork that might be a bit more controversial (like replacing underscores with hyphens in the flags).

martinjm97 commented:
Great to hear you're liking and using Tap!

Feel free to send over controversial PRs as well. @swansonk14 and I will have a few extended development sessions in about 2 weeks and we would definitely enjoy having your ideas in the mix!

martinjm97 commented:
Kyle and I expanded upon your PR and added support for primitives (including `int`, `float`, and `bool`) within literals among a number of other small changes. We committed it to your branch, but could not push. Any chance that you would be able to enable pushes from maintainers so that we can push and then merge into master (see https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)? If you could also do this for your other PRs it would be greatly appreciated. Thanks!

tmke8 commented:
Sounds great!

As far as I can tell, edits from maintainers are enabled. Maybe the problem was that the PRs were marked as "draft". I disabled the draft status now. Let me know in case it still doesn't work.
